numpy/lib/shape_base.py

```diff
@@ -16,9 +16,192 @@
 __all__ = [
     'column_stack', 'row_stack', 'dstack', 'array_split', 'split',
     'hsplit', 'vsplit', 'dsplit', 'apply_over_axes', 'expand_dims',
-    'apply_along_axis', 'kron', 'tile', 'get_array_wrap'
+    'apply_along_axis', 'kron', 'tile', 'get_array_wrap', 'take_along_axis',
+    'put_along_axis'
     ]
 
+def _make_along_axis_idx(arr, indices, axis):
+	# compute dimensions to iterate over
+    shape_ones = (1,) * indices.ndim
+    ins_ndim = indices.ndim - (arr.ndim - 1)  # inserted dimensions
+    if ins_ndim < 0:
+        raise ValueError("`indices` must have ndim >= arr.ndim - 1")
+    dest_dims = list(range(axis)) + [None] + list(range(axis+ins_ndim, indices.ndim))
+
+    # build a fancy index, consisting of orthogonal aranges, with the
+    # requested index inserted at the right location
+    fancy_index = []
+    for dim, n in zip(dest_dims, arr.shape):
+        if dim is None:
+            fancy_index.append(indices)
+        else:
+            ind_shape = shape_ones[:dim] + (-1,) + shape_ones[dim+1:]
+            fancy_index.append(_nx.arange(n).reshape(ind_shape))
+
+    return tuple(fancy_index)
+
+
+def take_along_axis(arr, indices, axis):
+    """
+    Take the elements described by `indices` along each 1-D slice of the given
+    `axis`, matching up subspaces of arr and indices.
+
+    This function can be used to index with the result of `argsort`, `argmax`,
+    and other `arg` functions.
+
+    For each `i...`, `j...`, `k...` in turn (representing a series of indices),
+    where the number of indices in `i...` is equal to `axis`, this computes::
+
+        out[i..., j..., k...] = arr[i..., indices[i..., j..., k...], k...]
+
+    Or equivalently (where `...` alone is the builtin `Ellipsis`):
+
+        out[i..., ..., k...] = arr[i..., :, k...][indices[i..., ..., k...]]
```

Comment:
Based on the discussion #9946, I suppose this would be better described as

````
Ni, Nk = a.shape[:axis], a.shape[axis+1:]
for ii in ndindex(Ni):
    for kk in ndindex(Nj):
        out[ii + s_[...,] + kk] = a[ii + s_[:,] + kk][indices[ii + s_[...,] + kk]]
```